pkg/scanners/trivy/trivy.go

```diff
@@ -113,170 +104,53 @@ var (
-	trivylogger.Logger = sugar
-
-	if err := unix.Mkfifo(util.EraseCompleteScanPath, util.PipeMode); err != nil {
-		log.Error(err, "failed to create pipe", "pipeName", util.EraseCompleteScanPath)
-		os.Exit(1)
-	}
-
-	err = os.Chmod(util.EraseCompleteScanPath, 0o666)
-	if err != nil {
-		log.Error(err, "unable to enable pipe for writing", "pipeName", util.EraseCompleteScanPath)
-		os.Exit(1)
-	}
-
 	if *enableProfile {
-		go func() {
-			server := &http.Server{
-				Addr:              fmt.Sprintf("localhost:%d", *profilePort),
-				ReadHeaderTimeout: 3 * time.Second,
-			}
-			err := server.ListenAndServe()
-			log.Error(err, "pprof server failed")
-		}()
+		go runProfileServer()
 	}
 
-	allSetsOfCommaSeparatedOptions := []optionSet{
-		{
-			input: *severity,
-			m:     severityMap,
-		},
-		{
-			input: *vulnTypes,
-			m:     vulnTypeMap,
-		},
-		{
-			input: *securityChecks,
-			m:     securityCheckMap,
-		},
-	}
-
-	for _, oSet := range allSetsOfCommaSeparatedOptions {
-		// note: this function has side effects and will modify the map supplied as the first argument
-		err := parseCommaSeparatedOptions(oSet.m, oSet.input)
-		if err != nil {
-			log.Error(err, "unable to parse options")
-			os.Exit(generalErr)
-		}
-	}
-
-	allImages, err := util.ReadCollectScanPipe(ctx)
-	if err != nil {
-		log.Error(err, "unable to read images from collect scan pipe")
-		os.Exit(generalErr)
-	}
-
-	err = downloadAndInitDB(*cacheDir)
+	ctx := context.Background()
+	provider := template.NewImageProvider(
+		template.WithContext(ctx),
+		template.WithLogger(log),
+		template.WithMetrics(true),
+		template.WithDeleteScanFailedImages(*deleteScanFailedImages),
+	)
+
+	allImages, err := provider.ReceiveImages()
 	if err != nil {
-		log.Error(err, "unable to initialize trivy db", "cacheDir", *cacheDir)
+		log.Error(err, "unable to read images from provider")
 		os.Exit(generalErr)
 	}
 
-	vulnTypeList := trueMapKeys(vulnTypeMap)
-	securityCheckList := trueMapKeys(securityCheckMap)
-
-	scanConfig, err := setupScanner(*cacheDir, vulnTypeList, securityCheckList)
+	s, err := initScanner(ctx)
 	if err != nil {
-		log.Error(err, "unable to set up scanner configuration", "cacheDir", *cacheDir, "vulnTypeList", vulnTypeList, "securityCheckList", securityCheckList)
-		os.Exit(generalErr)
+		log.Error(err, "error initializing scanner")
 	}
 
 	vulnerableImages := make([]eraserv1alpha1.Image, 0, len(allImages))
 	failedImages := make([]eraserv1alpha1.Image, 0, len(allImages))
 
-	runtime := os.Getenv(util.EnvEraserContainerRuntime)
-	imageSourceOptions, ok := runtimeFanalOptionsMap[runtime]
-	if !ok {
-		log.Error(err, "unable to determine runtime from environment")
-		os.Exit(generalErr)
-	}
-
 	for _, img := range allImages {
-		refs := make([]string, 0, len(img.Names)+len(img.Digests))
-		refs = append(refs, img.Digests...)
-		refs = append(refs, img.Names...)
-
-		if len(refs) == 0 {
-			log.Info("found image with no name", "img", img)
+		// Logs scan failures
```

Comment:
do you think we should put this in a scan() function to be consistent with the template?

Reply:
https://github.com/Azure/eraser-scanner-template/pull/1